source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -756,12 +776,22 @@ Network::FilterStatus Filter::onData(Buffer::Instance& data, bool end_stream) {
   if (upstream_) {
     getStreamInfo().getUpstreamBytesMeter()->addWireBytesSent(data.length());
     upstream_->encodeData(data, end_stream);
+    resetIdleTimer(); // TODO(ggreenway) PERF: do we need to reset timer on both send and receive?
+  } else if (receive_before_connect_) {
+    // Buffer data received before upstream connection exists
+    ENVOY_CONN_LOG(debug, "buffering {} bytes as upstream connection is not established yet",
+                   read_callbacks_->connection(), data.length());
+    early_data_buffer_.move(data);
+    read_callbacks_->connection().readDisable(true);
```

Comment:
This is to implement the flow control for the early_data_buffer, to prevent it from overflow. It addresses the feedback shared [here](https://github.com/envoyproxy/envoy/pull/25804#discussion_r1176786942)